Add exceptions support

mstemm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. . Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

/kind rule-update

> /kind rule-create

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

/area rules

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

This adds support for exceptions as rule attributes as described in https://github.com/falcosecurity/falco/pull/1376. Also included are a set of changes to falco_rules.yaml to use exceptions whenever possible.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
new: Add support for exceptions as rule attributes to provide a compact way to add exceptions to rules.
```

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [6480be8](https://github.com/falcosecurity/falco/commits/6480be85ce9a354abfc99695c0d7118d35b49ba9) Update tests to add error counts
- [3dd4f11](https://github.com/falcosecurity/falco/commits/3dd4f11862431cbbc1e667a44db5fe376158e15a) Support rules exceptions as standalone objects

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **mstemm**
You can assign the PR to them by writing `/assign @mstemm` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [7afa744](https://github.com/falcosecurity/falco/commits/7afa744d3752d622eabdf8448b3f5184cba8a74e) Squash w/ tests commit
- [d5b8d39](https://github.com/falcosecurity/falco/commits/d5b8d39f77b8420fee3ed0f2a8126d37c4779cd8) Squash w/ exceptions code commit

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [9cb25be](https://github.com/falcosecurity/falco/commits/9cb25be5bdeca6591da3b2d621bb3eb5adc16b56) Squash w/ test commit.
- [9c70ae1](https://github.com/falcosecurity/falco/commits/9c70ae19bea0b7a70c3322cdf23187be8f17eb12) Squash w/ code commit: single field exceptions

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [9cb25be](https://github.com/falcosecurity/falco/commits/9cb25be5bdeca6591da3b2d621bb3eb5adc16b56) Squash w/ test commit.
- [9c70ae1](https://github.com/falcosecurity/falco/commits/9c70ae19bea0b7a70c3322cdf23187be8f17eb12) Squash w/ code commit: single field exceptions
- [dbd4ff0](https://github.com/falcosecurity/falco/commits/dbd4ff08eb88271952e3645c6f12891a1d0e13e2) Rules changes (WIP)

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [9cb25be](https://github.com/falcosecurity/falco/commits/9cb25be5bdeca6591da3b2d621bb3eb5adc16b56) Squash w/ test commit.
- [9c70ae1](https://github.com/falcosecurity/falco/commits/9c70ae19bea0b7a70c3322cdf23187be8f17eb12) Squash w/ code commit: single field exceptions
- [dbd4ff0](https://github.com/falcosecurity/falco/commits/dbd4ff08eb88271952e3645c6f12891a1d0e13e2) Rules changes (WIP)
- [5e462f1](https://github.com/falcosecurity/falco/commits/5e462f12bb56f8a6361bf488f87e305f9e16b0ee) Addl rule loading changes/bug fixes

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [9cb25be](https://github.com/falcosecurity/falco/commits/9cb25be5bdeca6591da3b2d621bb3eb5adc16b56) Squash w/ test commit.
- [9c70ae1](https://github.com/falcosecurity/falco/commits/9c70ae19bea0b7a70c3322cdf23187be8f17eb12) Squash w/ code commit: single field exceptions
- [dbd4ff0](https://github.com/falcosecurity/falco/commits/dbd4ff08eb88271952e3645c6f12891a1d0e13e2) Rules changes (WIP)
- [5e462f1](https://github.com/falcosecurity/falco/commits/5e462f12bb56f8a6361bf488f87e305f9e16b0ee) Addl rule loading changes/bug fixes
- [f4297d0](https://github.com/falcosecurity/falco/commits/f4297d0a1b5cdbdb097c70cb9dd15ebca3bcadcf) More rule loader bug fixes
- [1d9a862](https://github.com/falcosecurity/falco/commits/1d9a8628228a4a3f458a38355d4cf09cb819288a) Add exceptions fields/comps/values to rules files

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
@mstemm: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [06257ed](https://github.com/falcosecurity/falco/commits/06257ed7af19b6dbfe0221792997d44c9c401912) Squash w/ Update tests to add error counts.
- [70df739](https://github.com/falcosecurity/falco/commits/70df7392bdac4ba95863148f526dd40e5ee9ae8c) Squash w/ rules change commit.
- [b4b0c29](https://github.com/falcosecurity/falco/commits/b4b0c29db4d3b610d808b3ebe76af2b85fd45a14) Squash w/ exceptions code commit.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

mstemm commented:
I believe this is ready for review now, with all the changes (including rules changes) in place and all the tests passing.

leogr reviewed (commented):
First of all, you did a great job :clap: 

I tried this new feature, and it seems to work very well.

Although it seems good to me, I need just a bit more time to complete the review since this PR is huge :smiley_cat:

Meanwhile, I have left just two comments (see below).

leogr commented:
Also, using some operators can produce an invalid YAML syntax, for example:
```yaml
comps: !=
```
or
```yaml
comps: [!=]
```

Not sure if that is a real issue anyway since it can be just solved using a duoble-quoted string.

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **mstemm** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @mstemm` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

mstemm commented on userspace/engine/lua/rule_loader.lua:
```diff
@@ -126,6 +126,31 @@ function set_output(output_format, state)
+defined_comp_operators = {
+   ["="]=1,
+   ["=="] = 1,
+   ["!"] = 1,
```
Yes, good catch! Fixed.

mstemm commented on userspace/engine/lua/rule_loader.lua:
```diff
@@ -592,52 +942,22 @@ function load_rules(sinsp_lua_parser,
 	 end
 
-	 evttypes, syscallnums = sinsp_rule_utils.get_evttypes_syscalls(name, filter_ast, v['condition'], warn_evttypes, verbose)
+	 evttypes, syscallnums = sinsp_rule_utils.get_evttypes_syscalls(name, filter_ast, v['compile_condition'], warn_evttypes, verbose)
```
I do think it's good to keep a verbose flag to suppress this message: https://github.com/falcosecurity/falco/blob/add-exceptions-support/userspace/engine/lua/sinsp_rule_utils.lua#L193. It can be useful when you're debugging, but I don't think you'd want to print it by default.

mstemm commented:
Okay, I rebased and fixed the one typo using !=. About the use of the operators direct in lists, I agree that the best solution will just be to quote the values. Also, I don't think the not equals style operators will be used that often anyway, as there's already an implied "not" by the use of the exception.

fntlnz commented:
/milestone 0.28.0
/hold

poiana commented:
LGTM label has been added.  <details>Git tree hash: 494c8d42685f23c0e17ea0a53d28f661458e84a3</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-563789452" title="Approved">fntlnz</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [66ad34e](https://github.com/falcosecurity/falco/commits/66ad34e931a01503d1a5e8a8655f9a60e283168e) Make the req. engine version 8 for k8s_audit rules

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

mstemm commented:
Added a couple of small changes to make it easier to deal with exceptions when processing events.

leodido commented on rules/falco_rules.yaml:
```diff
@@ -395,10 +381,20 @@
-     (fd.snet in (allowed_outbound_destination_networks)) or
-     (fd.sip.name in (allowed_outbound_destination_domains)))
+    consider_all_outbound_conns and outbound
+  exceptions:
```
Mark, just out of curiosity, have you run some benchmarks to evaluate the performance changes?

I mean, expressing "exceptions" with this new first-class citizen is faster or slower than expressing them with the "not" operator in the condition.

leodido replied:
I imagine there's little (negligible) overhead during the parsing phase.

I don't expect performance to decrease at runtime since exceptions get translated...

Am I correct?

mstemm replied:
The exceptions are folded into the condition string at the time the rules are loaded, so I'm not worried about the conversion overhead.

I think the resulting condition is as about as efficient as the current chain of macros are. Each macro effectively was one tuple of exception fields + values.

leodido replied:
Yeah, exactly what I was supposing. Thanks for confirming it Mark!

leodido commented on test/falco_tests.yaml:
```diff
@@ -292,20 +295,10 @@ trace_files: !mux
       - rules/invalid_array_item_not_object.yaml
     trace_file: trace_files/cat_write.scap
 
-  invalid_unexpected object:
```
Is this deletion intended?

leodido replied:
I saw that you deleted `test/rules/invalid_unexpected_object.yaml`. Is this test not needed anymore?

mstemm replied:
Yes, this test was removed as the rule loading is now tolerant of unknown top level objects. This will allow defining new top level objects (something analogous to rule/exception/macro/list) without interfering with older falco instances that might not know about them.

leodido replied:
Ok, that sounds like something to mention in the release notes :)

Because users may find ways to use this feature to annotate Falco rules ;)

leodido commented on userspace/engine/falco_engine.cpp:
```diff
@@ -350,33 +328,50 @@ unique_ptr<falco_engine::rule_result> falco_engine::process_k8s_audit_event(json
 	}
 
 	unique_ptr<struct rule_result> res(new rule_result());
+	res->source = "k8s_audit";
```
👍

leodido reviewed (commented):
Good job!

Also, the extensive test suite for exceptions is lovely!

Left just some comments, but this looks good to me (as long the current failing tests get fixed).

poiana commented:
@mstemm: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leodido reviewed (approved):
LGTM

But let's keep this on hold until next week (ie., after Falco 0.27.0 is released)

poiana commented:
LGTM label has been added.  <details>Git tree hash: 378709e3b88732e0c555a675780ae1322194e246</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-565199529" title="Approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [leodido]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-565200238" title="Approved">fntlnz</a>*, *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-565199529" title="Approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz,leodido]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-565200238" title="Approved">fntlnz</a>*, *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-565199529" title="Approved">leodido</a>*, *<a href="https://github.com/falcosecurity/falco/pull/1427#pullrequestreview-565201077" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz,leodido,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

fntlnz commented:
/hold cancel
